logutil/logutil.go

```diff
@@ -0,0 +1,132 @@
+// It implements io.Writer as well as a Set method, and is safe for use by multiple goroutines.
+//
+// A WriterVar must not be copied after first use.
+//
+// The zero value LevelVar is a no-op writer that discards all data written to it (similar to io.Discard).
+type WriterVar struct {
+	w  io.Writer
+	mu sync.Mutex
+}
+
+// NewWriterVar creates a new WriterVar with the given writer.
+func NewWriterVar(w io.Writer) *WriterVar {
+	return &WriterVar{w: w}
+}
+
+// Set sets the underlying writer.
+func (v *WriterVar) Set(w io.Writer) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	v.w = w
+}
+
+func (v *WriterVar) Write(p []byte) (int, error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	// Handle zero value case, if no writer just no-op.
+	if v.w == nil {
+		return len(p), nil
+	}
+	return v.w.Write(p)
+}
+
+// LogWriter returns an io.Writer that can be used to write arbitrary text to the logger.
+// logger is used to log an error if one occurs.
+//
+// It is the caller's responsibility to close the returned io.WriteCloser in order
+// to free resources.
+func LogWriter(logger progress.Logger, level slog.Level) io.WriteCloser {
+	pr, pw := io.Pipe()
+	var logFunc func(string, ...any)
+	switch level {
+	case slog.LevelDebug:
+		logFunc = logger.Debug
+	case slog.LevelInfo:
+		logFunc = logger.Info
+	case slog.LevelWarn:
+		logFunc = logger.Warn
+	case slog.LevelError:
+		logFunc = logger.Error
+	default:
+		// See if the logger has a Log method that can be passed a level.
+		type withlog interface {
+			Log(context.Context, slog.Level, string, ...any)
+		}
+		wl, ok := logger.(withlog)
+		if !ok {
+			panic(fmt.Errorf("logutil.LogWriter: unsupported level %s(%d)", level.String(), level))
+		}
+		logFunc = func(s string, a ...any) {
+			wl.Log(context.Background(), level, s, a...)
+		}
+	}
+	go logText(logger, pr, logFunc)
+	runtime.SetFinalizer(pw, (*io.PipeWriter).Close)
+	return pw
+}
+
+func logText(logger progress.Logger, pr *io.PipeReader, logFunc func(string, ...any)) {
+	s := bufio.NewScanner(pr)
+	for s.Scan() {
+		logFunc(s.Text())
+	}
+	if err := s.Err(); err != nil {
+		logger.Error("Error while reading from Writer", "err", err)
+	}
+	pr.Close()
+}
+
+// CallerPC returns the program counter at the given stack depth.
+func CallerPC(depth int) uintptr {
+	var pcs [1]uintptr
+	// Need to add +1 to depth in order to skip this function.
+	runtime.Callers(depth+1, pcs[:])
+	return pcs[0]
+}
+
+// CallerSource returns a slog.Source for the given program counter.
+// If the location is unavailable, it returns a slog.Source with zero fields.
+func CallerSource(pc uintptr) slog.Source {
+	fs := runtime.CallersFrames([]uintptr{pc})
+	f, _ := fs.Next()
+	return slog.Source{
+		Function: f.Function,
+		File:     f.File,
+		Line:     f.Line,
+	}
+}
+
+// RemoveKeys returns a function suitable for HandlerOptions.ReplaceAttr
+// that removes all Attrs with the given keys.
+func RemoveKeys(keys ...string) func([]string, slog.Attr) slog.Attr {
```

Comment:
Doubt this is a big deal but one thing we could do is build a map in the closure's environment, and have the func read from that map in O(1). I'll have to look ahead and see when this is called to see if it makes sense

Reply:
I don't think it's a big deal in practice, I don't expect the list of `keys` to be that large. Looking over a slice that is small shouldn't cause that big of an overhead, I wonder what the tradeoff is there vs the memory required to hold the map.

E.g. here's how it would be used in tb to remove the time key from logs
```go
logutil.NewPrettyHandler(w, &logutil.PrettyHandlerOptions{
	Level:       level,
	ReplaceAttr: logutil.RemoveKeys(slog.TimeKey),
}),
```

I can change it, would be easy, just wondering if it's worth it.